confluent_kafka/avro/cached_schema_registry_client.py

```diff
@@ -0,0 +1,326 @@
+        if body:
+            body = json.dumps(body)
+            body = body.encode('utf8')
+        _headers = dict()
+        _headers["Accept"] = ACCEPT_HDR
+        if body:
+            _headers["Content-Length"] = str(len(body))
+            _headers["Content-Type"] = "application/vnd.schemaregistry.v1+json"
+
+        if headers:
+            for header_name in headers:
+                _headers[header_name] = headers[header_name]
+        if method == 'GET':
+            response = requests.get(url, headers=_headers)
+        elif method == 'POST':
+            response = requests.post(url, body, headers=_headers)
+        elif method == 'PUT':
+            response = requests.put(url, body, headers=_headers)
+        elif method == 'DELETE':
+            response = requests.delete(url, headers=_headers)
+        result = json.loads(response.text)
+        return (result, response.status_code)
+
+    def _add_to_cache(self, cache, subject, schema, value):
+        sub_cache = cache[subject]
+        sub_cache[schema] = value
+
+    def _cache_schema(self, schema, schema_id, subject=None, version=None):
+        # don't overwrite anything
+        if schema_id in self.id_to_schema:
+            schema = self.id_to_schema[schema_id]
+        else:
+            self.id_to_schema[schema_id] = schema
+
+        if subject:
+            self._add_to_cache(self.subject_to_schema_ids,
+                               subject, schema, schema_id)
+            if version:
+                self._add_to_cache(self.subject_to_schema_versions,
+                                   subject, schema, version)
+
+    def register(self, subject, avro_schema):
+        """
+        POST /subjects/(string: subject)/versions
+        Register a schema with the registry under the given subject
+        and receive a schema id.
+
+        avro_schema must be a parsed schema from the python avro library
+
+        Multiple instances of the same schema will result in cache misses.
+
+        @:param: subject: subject name
+        @:param: avro_schema: Avro schema to be registered
+        @:returns: schema_id: int value
+        """
+
+        schemas_to_id = self.subject_to_schema_ids[subject]
+        schema_id = schemas_to_id.get(avro_schema, None)
+        if schema_id != None:
+            return schema_id
+        # send it up
+        url = '/'.join([self.url, 'subjects', subject, 'versions'])
+        # body is { schema : json_string }
+
+        body = {'schema': json.dumps(avro_schema.to_json())}
+        result, code = self._send_request(url, method='POST', body=body)
+        if code == 409:
+            log.error("Incompatible Avro schema:" + str(code))
+            raise ClientError("Incompatible Avro schema:" + str(code))
+        elif code == 422:
+            log.error("Invalid Avro schema:" + str(code))
+            raise ClientError("Invalid Avro schema:" + str(code))
+        elif code != 200:
```

Comment:
I think I mentioned this elsewhere, but every status in the 200s is valid, so this should probably just be `elif code >= 300` (and maybe < 200).